File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include <bits/stdc++.h>
using namespace std;

bool solution(string s) {
    ios::sync_with_stdio(0); cin.tie(0);
    
    stack<char> st;
    
    for (char c : s) {
        if (c == '(') {
            st.push(c);
        } 
        else if (c == ')') {
            if (st.empty()) {
                return false;
            }
            else if (st.top() == '(') {
                st.pop();
            }
        }
    }
    return st.empty();
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
class Solution {
    boolean solution(String s) {
        int n = 0;
        for (char c : s.toCharArray()) {
            if(n < 0) {
                return false;
            } else if (c == '(') {
                ++n;
            } else if (c == ')') {
                --n;
            }
        }

        return (n == 0);
    }
}